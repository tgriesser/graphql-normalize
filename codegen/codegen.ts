import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    //
  },
};

export default NormalizeCodegenPlugin;
