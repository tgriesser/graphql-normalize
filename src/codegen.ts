import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
