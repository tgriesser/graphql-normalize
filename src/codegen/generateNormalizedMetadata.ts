import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer'
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  OperationTypeNode,
  isObjectType,
  isScalarType,
  InlineFragmentNode,
  getNamedType,
  assertObjectType,
  isEnumType,
} from 'graphql'
import type {
  FieldDef,
  FieldMeta,
  NormalizeMetaShape,
  NormalizedDocShape,
  UnionMeta,
  VariableMeta,
} from '../metadataShapes'
import { TypePolicies, getCacheKey } from './getCacheKey'
import { stringifyVariables } from '../stringifyVariables'

export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedMetadataForDocs(schema, [document], typePolicies)[0].meta
}

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadataForDocs(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
): NormalizedDocShape[] {
  let finalDocuments = documents
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    finalDocuments = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    })
  }

  const normalizedDocs: NormalizedDocShape[] = []

  const typeInfo = new TypeInfo(schema)
  let normalizedDoc: NormalizeMetaShape = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    fields: [],
  }

  let parentStack: Array<FieldMeta | UnionMeta | NormalizeMetaShape> = []
  let parentFieldDef: FieldMeta | NormalizeMetaShape | UnionMeta = normalizedDoc

  function popStack() {
    const parent = parentStack.pop()
    if (parent) {
      parentFieldDef = parent
    }
  }

  function pushField(field: string | FieldDef) {
    parentFieldDef.fields ??= []
    if (typeof field === 'string' && parentFieldDef.fields.includes(field)) {
      return
    }
    parentFieldDef.fields.push(field)
    if (typeof field !== 'string') {
      parentStack.push(parentFieldDef)
      parentFieldDef = field
    }
  }

  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false
    const argsDef: Record<string, any> = {}
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.value.name.value}`] = arg.name.value
        hasVar = true
      } else {
        argsDef[arg.name.value] = print(arg.value)
      }
    }
    return hasVar ? argsDef : stringifyVariables(argsDef)
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition: {
      enter(op) {
        normalizedDoc.operation = op.operation
      },
      leave(op) {
        normalizedDocs.push({
          name: op.name?.value ?? 'unnamed',
          meta: normalizedDoc,
        })
        normalizedDoc = {
          operation: OperationTypeNode.QUERY,
          variables: [],
          fields: [],
        }
        parentFieldDef = normalizedDoc
        parentStack = []
      },
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      }
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue)
      }
      normalizedDoc.variables.push(obj)
    },
    Field: {
      enter(node) {
        const { gqlType, listDepth } = unpackType(typeInfo)

        // If this is a boring ol scalar / enum, we push it as a string
        if (isSimpleField(node, typeInfo)) {
          pushField(node.name.value)
          return
        }

        const fieldMeta: FieldMeta = { name: node.name.value }
        pushField(fieldMeta)

        if (listDepth) fieldMeta.list = listDepth
        if (node.alias?.value) fieldMeta.alias = node.alias.value

        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType)
          if (cacheKey) {
            fieldMeta.cacheKey = cacheKey
          }
          if (gqlType.name === schema.getQueryType()?.name) {
            fieldMeta.cacheKey = '$root'
          }
        }
        if (node.arguments?.length) {
          fieldMeta.args = normalizeArgs(node.arguments)
        }

        const skipDirective = node.directives?.find((d) => d.name.value === 'skip')
        const includeDirective = node.directives?.find((d) => d.name.value === 'include')
        if (skipDirective) {
          fieldMeta.skip = normalizeArgs(skipDirective.arguments ?? [])
        }
        if (includeDirective) {
          fieldMeta.include = normalizeArgs(includeDirective.arguments ?? [])
        }
      },
      leave(node) {
        // Pop things back into place
        if (!isSimpleField(node, typeInfo)) {
          popStack()
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (
          node.typeCondition?.name.value &&
          isConcreteAbstract(node, typeInfo, schema) &&
          parentIsFieldMeta(parentFieldDef)
        ) {
          const unionType: UnionMeta = {
            fields: [],
          }
          const outputType = getNamedType(typeInfo.getType())
          const cacheKey = getCacheKey(typePolicies, assertObjectType(outputType))
          if (cacheKey) {
            unionType.cacheKey = cacheKey
          }

          parentFieldDef.possible ??= {}
          parentFieldDef.possible[node.typeCondition?.name.value] = unionType
          parentStack.push(parentFieldDef)
          parentFieldDef = unionType
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value && isConcreteAbstract(node, typeInfo, schema)) {
          popStack()
        }
      },
    },
    FragmentDefinition() {
      throw noFragments()
    },
  })

  for (const doc of finalDocuments) {
    visit(doc, visitor)
  }

  return normalizedDocs
}

function unpackType(typeInfo: TypeInfo) {
  let listDepth = 0
  // We want to unwrap any list types, and make those the containing
  let gqlType = typeInfo.getType()
  while (isWrappingType(gqlType)) {
    if (isListType(gqlType)) {
      ++listDepth
    }
    gqlType = gqlType.ofType
  }
  return { gqlType, listDepth }
}

function isSimpleField(node: FieldNode, typeInfo: TypeInfo) {
  const { gqlType, listDepth } = unpackType(typeInfo)
  return (
    (isScalarType(gqlType) || isEnumType(gqlType)) &&
    !listDepth &&
    !node.alias &&
    !node.arguments?.length &&
    !node.directives?.some((d) => d.name.value === 'skip' || d.name.value === 'includes')
  )
}

function noFragments() {
  return new Error('Fragment definitions have been stripped')
}

function isConcreteAbstract(node: InlineFragmentNode, typeInfo: TypeInfo, schema: GraphQLSchema) {
  if (node.typeCondition?.name) {
    const parentTypeName = typeInfo.getParentType()?.name
    if (node.typeCondition.name.value !== parentTypeName) {
      return isObjectType(schema.getType(node.typeCondition?.name.value))
    }
  }
  return false
}

function parentIsFieldMeta(parent: FieldMeta | NormalizeMetaShape | UnionMeta): parent is FieldMeta {
  return 'name' in parent
}
