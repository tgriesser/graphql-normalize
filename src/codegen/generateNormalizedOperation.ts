import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';
import type { TypePolicies } from './getCacheKey';

export function generateNormalizedOperation(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedOperations(schema, [document], typePolicies)[0];
}

/**
 * Given a schema & operation, flattens the fragments and adds any necessary missing fields
 * in order to properly normalize the response in a cache.
 */
export function generateNormalizedOperations(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
) {
  let toNormalize = documents;
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    toNormalize = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    });
  }
  return addNormalizingFields(schema, toNormalize, typePolicies);
}
