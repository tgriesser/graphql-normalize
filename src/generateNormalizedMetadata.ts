import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isAbstractType,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  SelectionSetNode,
  OperationTypeNode,
  isObjectType,
} from 'graphql';
import stableStringify from 'fast-json-stable-stringify';
import type { FieldDef, FieldMeta, NormalizeMeta, NormalizedDoc, VariableMeta } from './extensionShape';
import { TypePolicies, getCacheKey } from './typePolicies';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizedDoc {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizedDoc = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    selectionSet: {},
  };

  const fieldPath: any[] = [];

  function setPath<K extends keyof NormalizeMeta>(key: K, value: NormalizeMeta[K]) {
    normalizedDoc.selectionSet[fieldPath.join('.')] ??= {};
    normalizedDoc.selectionSet[fieldPath.join('.')][key] = value;
  }
  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stableStringify(argsDef);
  }
  function normalizeField(node: FieldNode) {
    if (node.alias || node.arguments?.length) {
      const field: FieldMeta = {
        name: node.name.value,
      };
      if (node.alias) field.alias = node.alias.value;
      if (node.arguments?.length) field.args = normalizeArgs(node.arguments);
      return field;
    }
    return node.name.value;
  }
  function addSelectionFields(node: SelectionSetNode) {
    const fields: FieldDef[] = [];
    for (const sel of node.selections) {
      if (sel.kind === Kind.FRAGMENT_SPREAD) {
        throw noFragments();
      }
      if (sel.kind === Kind.FIELD) {
        fields.push(normalizeField(sel));
      }
      // If we have an inline fragment
      if (sel.kind === Kind.INLINE_FRAGMENT) {
        //
      }
    }
    setPath('fields', fields);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        if (node.alias?.value) {
          fieldPath.push(node.alias.value);
          setPath('aliasField', node.name.value);
        } else {
          fieldPath.push(node.name.value);
        }
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            setPath('list', true);
            fieldPath.push('$idx');
          }
          gqlType = gqlType.ofType;
        }
        if (isAbstractType(gqlType)) {
          setPath(
            'possible',
            schema.getPossibleTypes(gqlType).map((t) => t.name)
          );
        }
        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            setPath('cacheKey', cacheKey);
          }
        }
        if (node.arguments?.length) {
          setPath('args', normalizeArgs(node.arguments));
        }
        if (node.selectionSet) {
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        fieldPath.pop();
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            fieldPath.pop();
          }
          gqlType = gqlType.ofType;
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (node.typeCondition?.name) {
          fieldPath.push(`$type_${node.typeCondition?.name.value}`);
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value) {
          fieldPath.pop();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}
