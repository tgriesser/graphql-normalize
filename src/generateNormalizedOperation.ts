import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';

export function generateNormalizedOperation(schema: GraphQLSchema, operation: DocumentNode) {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }
  return addNormalizingFields(schema, finalOperation);
}
