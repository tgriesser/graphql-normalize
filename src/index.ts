import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
