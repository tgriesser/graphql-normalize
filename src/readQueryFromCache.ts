import type { Immutable } from 'immer';
import type { CacheShape } from './cache';
import type { FieldDef, FieldMeta, NormalizeMetaShape } from './metadataShapes';
import { NO_ARGS, printArgs } from './printArgs';
import { copyList } from './copyList';

interface ReadQueryFromCache<Q> {
  meta: NormalizeMetaShape;
  cacheFields: Immutable<CacheShape['fields']>;
  variableValues: Record<string, any>;
  existing?: Q;
  isEqual?: (a: any, b: any) => boolean;
}

export class MissingFieldError {
  constructor(path: Array<string | number>) {
    throw new Error(`Missing field for: ${path.filter((p) => p !== NO_ARGS)}`);
  }
}

const defaultIsEqual = (a: any, b: any) => {
  return a === b;
};

interface TraverseFieldsConfig {
  into: any;
  existing: any;
  fields: FieldDef[];
  basePath: Array<string | number>;
}

/**
 * Reads a query from the cache. We do this in order to ensure we're only
 * providing the fields the original query has requested, rather than
 * any additional fields fetched in the process of normalization
 */
export function readQueryFromCache<Q>(config: ReadQueryFromCache<Q>): Q {
  const { meta, cacheFields, variableValues = {}, existing = {} as Q, isEqual = defaultIsEqual } = config;
  const { fields, operation, variables } = config.meta;

  function shouldSkip(field: FieldMeta) {
    return false;
  }

  function traverseFields(config: TraverseFieldsConfig): any {
    const { fields, into = {}, existing, basePath } = config;
    let changed = false;

    const getField = (path: Array<string | number>) => {
      return basePath.concat(path).reduce((acc, key) => acc?.[key], cacheFields);
    };

    const getValue = (path: Array<string | number>) => {
      return path.reduce((acc, key) => acc?.[key], existing);
    };

    const setField = (key: string, value: any) => {
      if (Array.isArray(into) && typeof key !== 'number') {
        // console.log({ into, key, basePath });
      }

      if (into[key] === undefined || !isEqual(into[key], value)) {
        into[key] = value;
        changed = true;
      }
    };

    for (const field of fields) {
      if (field === '__typename') {
        setField(field, getField(['__typename']));
      } else if (typeof field === 'string') {
        setField(field, getField([field, NO_ARGS]));
      } else {
        const argsKey = printArgs(field.args, meta, variableValues);
        const resultName = field.alias ?? field.name;
        const fieldData = getField([field.name, argsKey]);

        if (shouldSkip(field)) {
          if (resultName in into) {
            delete into[resultName];
            changed = true;
            continue;
          }
        }

        // If we have a null value, there's nothing more to do here
        if (fieldData === null) {
          setField(resultName, null);
          continue;
        }

        const handleField = (indexes: number[] = []) => {
          const cachedValue = getField([field.name, argsKey, ...indexes]);
          const fieldPath = field.cacheKey
            ? [cachedValue.$ref]
            : basePath.concat([field.name, argsKey]).concat(indexes);

          if (field.fields) {
            const { changed, result } = traverseFields({
              into: getValue([resultName, ...indexes]) ?? {},
              existing: existing?.[resultName],
              fields: field.fields,
              basePath: fieldPath,
            });

            return result;
          }
        };

        // If we know we have a list, copy items at the depth of the list
        if (field.list) {
          const { list, modified } = copyList({
            isEqual,
            source: Array.isArray(fieldData) ? fieldData : [],
            depth: field.list,
            target: into[resultName] ?? [],
            onValue: (existingValue, newValue, indexes) => handleField(indexes),
          });
          if (modified) {
            setField(resultName, list);
          }
        } else {
          into[resultName] = handleField();
        }
      }
    }

    return { changed, result: into };
  }

  const { changed, result } = traverseFields({
    into: existing,
    existing,
    fields,
    basePath: [],
  });

  return changed ? result : existing;
}
