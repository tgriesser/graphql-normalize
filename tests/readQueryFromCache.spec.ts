import { execute } from 'graphql';
import { beforeEach, describe, expect, it } from 'vitest';
import { produce } from 'immer';

import { generateNormalizedOperation } from '../src/codegen/generateNormalizedOperation';
import { schema } from './fixtures/schema';
import { operation1Doc } from './fixtures/ops';
import { generateNormalizedMetadata } from '../src/codegen/generateNormalizedMetadata';
import { readQueryFromCache } from '../src/readQueryFromCache';
import type { NormalizeMetaShape } from '../src/metadataShapes';
import { syncWithCache } from '../src/syncWithCache';

describe('readQueryFromCache', () => {
  let cache = {
    operations: {},
    fields: {},
  };
  let meta: NormalizeMetaShape;
  beforeEach(async () => {
    meta = generateNormalizedMetadata(schema, operation1Doc);
    const variableValues = {};
    const result = await execute({
      schema,
      variableValues,
      document: generateNormalizedOperation(schema, operation1Doc),
    });
    cache = {
      operations: {},
      fields: {},
    };
    syncWithCache({
      action: 'write',
      meta,
      operationResult: result.data,
      variableValues,
      cache,
    });
  });

  it('builds a cache of values', async () => {
    const readResult = readQueryFromCache({
      meta,
      cacheFields: cache.fields,
      variableValues: {},
    });
    expect(readResult).toMatchSnapshot();
  });

  it('does not mutate the data if it has not changed', async () => {
    const readResult = readQueryFromCache({
      meta,
      cacheFields: cache.fields,
      variableValues: {},
    });
    const initial = produce({}, () => readResult);
    const secondRead = produce(initial, (existing) => {
      return readQueryFromCache({
        meta,
        cacheFields: cache.fields,
        existing,
        variableValues: {},
      });
    });
    expect(secondRead).toEqual(initial);
  });
});
