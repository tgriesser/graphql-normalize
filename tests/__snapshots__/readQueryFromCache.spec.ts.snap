// Vitest Snapshot v1

exports[`readQueryFromCache > builds a cache of values 1`] = `
{
  "currentNode": undefined,
  "first10Posts": {
    "nodes": [
      {
        "id": "UG9zdDox",
        "title": "Some Blog post 1",
      },
      {
        "id": "UG9zdDoy",
        "title": "Some Blog post 2",
      },
      {
        "id": "UG9zdDoz",
        "title": "Some Blog post 3",
      },
      {
        "id": "UG9zdDo0",
        "title": "Some Blog post 4",
      },
      {
        "id": "UG9zdDo1",
        "title": "Some Blog post 5",
      },
      {
        "id": "UG9zdDo2",
        "title": "Some Blog post 6",
      },
      {
        "id": "UG9zdDo3",
        "title": "Some Blog post 7",
      },
      {
        "id": "UG9zdDo4",
        "title": "Some Blog post 8",
      },
      {
        "id": "UG9zdDo5",
        "title": "Some Blog post 9",
      },
      {
        "id": "UG9zdDoxMA==",
        "title": "Some Blog post 10",
      },
    ],
  },
  "homeItems": {},
  "nestedUserList": [
    [
      {
        "name": "user 1",
      },
    ],
    [
      {
        "name": "user 2",
      },
    ],
  ],
  "num": 1,
  "posts": {
    "edges": [
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDox",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoy",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoz",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo0",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo1",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo2",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo3",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo4",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo5",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMA==",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoxMQ==",
        },
      },
      {
        "node": {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMg==",
        },
      },
    ],
    "nodes": [
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDox",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoy",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDoz",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo0",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo1",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo2",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo3",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDo4",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDo5",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoxMA==",
      },
      {
        "author": {
          "id": "VXNlcjox",
        },
        "id": "UG9zdDoxMQ==",
      },
      {
        "author": {
          "id": "VXNlcjoy",
        },
        "id": "UG9zdDoxMg==",
      },
    ],
  },
}
`;
